libs/analysis/bbhnet/analysis/vt.py

```diff
@@ -0,0 +1,166 @@
+import math
+from dataclasses import dataclass
+from typing import TYPE_CHECKING, Dict, Optional
+
+from astropy import cosmology as cosmo
+from astropy import units as u
+
+if TYPE_CHECKING:
+    from astropy.quantity import Quantity
+    from astropy.cosmology import Cosmology
+    import bilby
+
+import numpy as np
+from scipy.integrate import quad
+
+PI_OVER_TWO = math.pi / 2
+YEARS_PER_SECOND = 1 / (60 * 60 * 24 * 365)
+
+
+def calculate_astrophysical_volume(
+    dl_min: float,
+    dl_max: float,
+    dec_min: float = -PI_OVER_TWO,
+    dec_max: float = PI_OVER_TWO,
+    cosmology: "Cosmology" = cosmo.Planck15,
+) -> "Quantity":
+    """
+    Calculates the astrophysical volume over which injections have been made.
+    See equation 4) in https://arxiv.org/pdf/1712.00482.pdf
+
+    Args:
+        dl_min: minimum distance of injections in Mpc
+        dl_max: maximum distance of injections in Mpc
+        dec_min: minimum declination of injections in radians
+        dec_max: maximum declination of injections in radians
+        cosmology: astropy cosmology object
+
+    Returns astropy.Quantity of volume in Mpc^3
+    """
+
+    # given passed cosmology, calculate the comoving distance
+    # at the minimum and maximum distance of injections
+    zmin, zmax = cosmo.z_at_value(
+        cosmology.luminosity_distance, [dl_min, dl_max] * u.Mpc
+    )
+
+    # calculate the angular volume of the sky
+    # over which injections have been made
+    dec_min, dec_max = dec_min * u.rad, dec_max * u.rad
+    theta_max, theta_min = (
+        np.pi / 2 - dec_min.value,
+        np.pi / 2 - dec_max.value,
+    )
+    omega = -2 * math.pi * (np.cos(theta_max) - np.cos(theta_min))
+
+    # calculate the volume of the universe
+    # over which injections have been made
+    integrand = (
+        lambda z: 1.0 * (cosmology.differential_comoving_volume(z)).value
```

Comment:
We had a brief chat about this. Turns out the `UniformInComovingVolume` in bilby does not have the time dilation factor. What they call `UniformInSourceFrame` is what is uniform in comoving volume typically in GW literature. See: https://git.ligo.org/lscsoft/bilby/-/blob/master/bilby/gw/prior.py#L311-346

Hence, let's keep the prior for distance as uniform in source frame, but use the 1/(1 + z) factor here.

```suggestion
        lambda z: 1 / (1 + z) * (cosmology.differential_comoving_volume(z)).value
```